src/comparator/message_comparator.py

```diff
@@ -86,5 +96,86 @@ def _compareNestedMessages(self, nested_msg_dict_original, nested_msg_dict_updat
+        # the resource could be moved to file-level resource definition.
+        # 3. Note that the type change of an existing resource definition is like one resource
+        # is removed and another one is added.
+        if (resource_original and not resource_update) or (
+            resource_original.type != resource_update.type
+        ):
+            # Check if the removed resource is in the global file-level resource database.
+            if resource_original.type not in self.global_resources_update.types:
+                FindingContainer.addFinding(
+                    FindingCategory.RESOURCE_DEFINITION_REMOVAL,
+                    "",
+                    f"A message-level resource definition {resource_original.type} has been removed.",
+                    True,
+                )
+            else:
+                # Check the patterns of existing file-level resource are compatible with
+                # the patterns of the removed message-level resource.
+                global_resource_pattern = self.global_resources_update.types[
+                    resource_original.type
+                ].pattern
+
+                # If there is pattern removal, or pattern value change. Then the global file-level resource
+                # can not replace the original message-level resource.
+                if self._compatible_patterns(
+                    resource_original.pattern, global_resource_pattern
+                ):
+                    FindingContainer.addFinding(
+                        FindingCategory.RESOURCE_DEFINITION_REMOVAL,
+                        "",
+                        f"A message-level resource definition {resource_original.type} has been removed.",
+                        True,
+                    )
+            return
+        # Resource is existing in both original and update versions.
+        # 3. Patterns of message-level resource definitions have changed.
+        if self._compatible_patterns(
+            resource_original.pattern, resource_update.pattern
+        ):
+            FindingContainer.addFinding(
+                FindingCategory.RESOURCE_DEFINITION_CHANGE,
+                "",
+                f"The pattern of message-level resource definition has changed from {resource_original.pattern} to {resource_update.pattern}.",
+                True,
+            )
+
+    def _get_resource_option(self, message):
+        resource = message.options.Extensions[resource_pb2.resource]
+        if not resource.type or not resource.pattern:
+            return None
+        return resource
+
+    def _compatible_patterns(self, patterns_original, patterns_update):
+        # An existing pattern is removed.
+        if len(patterns_original) > len(patterns_update):
+            return True
+        # An existing pattern value is changed.
+        for old_pattern, new_pattern in zip(patterns_original, patterns_update):
```

Comment:
not sure what the types of `patterns_original` and `patterns_update` are. If they are list, just sort them, in case they are the same list but contents are in different order.

Reply:
Similar to the method_signatures, the order of patterns in a resource definition matters. So we probably want to compare the two list directly with position information. But thanks for the reminder!